### PR TITLE
Reland "[InstCombine] Fold (sub nuw X, (Y << nuw Z)) >>u exact Z --> (X >>u exact Z) sub nuw Y"

### PR DESCRIPTION
**Pull request by AZero13**
This is the same fold as ((X << nuw Z) sub nuw Y) >>u exact Z --> X sub nuw (Y >>u exact Z), but with the sub operands swapped.

Alive2 Proof:
https://alive2.llvm.org/ce/z/pT-RxG

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1280]: Please provide alive2 proof for nsw flag propagation.

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/lshr.ll:469]: Please add a test without nsw flag as well.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1280]: Done!

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/lshr.ll:469]: Done!

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1280]: Please double-check your proofs :)

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1280]: Done and done!

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1280]:
```
define i16 @src1(i16 %x, i16 noundef %c, i16 %y) {
  %shl = shl nuw nsw i16 %x, %c
  %sub = sub nuw i16 %y, %shl
  %lshr = lshr exact i16 %sub, %c
  ret i16 %lshr
}

define i16 @tgt1(i16 %x, i16 noundef %c, i16 %y) {
  %tmp1 = lshr exact i16 %y, %c
  %sub = sub nuw i16 %tmp1, %x
  ret i16 %sub
}
```

`nsw` here is unnecessary.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1280]: I fixed that.

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/lshr.ll:507]: Can you add more negative tests?
1. multi-use on sub
2. multi-use on shl
3. sub without nuw
4. shl without nuw

**Comment by AtariDreams** [comment]
@dtcxzyw Is this good to be relanded now?

File: llvm/test/Transforms/InstCombine/lshr.ll
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/lshr.ll:507]: Alright

**Comment by dtcxzyw** [comment]
BTW, don't @ someone in the PR description/commit message, it is annoying.

**Comment by AtariDreams** [comment]
@dtcxzyw I did all the possible negative tests. 

I think this can be merged now!

**Comment by AtariDreams** [comment]
@goldsteinn Any updates?

**Comment by AtariDreams** [comment]
@nikic Ping?

**Comment by goldsteinn** [comment]
This LGTM, please wait for another approval before pushing.

**Review by nikic — APPROVED** [review]
LGTM.

I'll merge this on Monday if nobody beats me to it.

**Pull request merged**
